swarm/src/behaviour.rs

```diff
@@ -31,26 +31,116 @@ use std::{task::Context, task::Poll};
 type THandlerInEvent<THandler> =
     <<THandler as IntoProtocolsHandler>::Handler as ProtocolsHandler>::InEvent;
 
-/// A behaviour for the network. Allows customizing the swarm.
+/// A [`NetworkBehaviour`] defines the behaviour of the local node on the network.
 ///
-/// This trait has been designed to be composable. Multiple implementations can be combined into
-/// one that handles all the behaviours at once.
+/// In contrast to [`Transport`](libp2p_core::Transport) which defines **how** to send bytes on the
+/// network, [`NetworkBehaviour`] defines **what** bytes to send and **to whom**.
 ///
-/// # Deriving `NetworkBehaviour`
+/// Each protocol (e.g. `libp2p-ping`, `libp2p-identify` or `libp2p-kad`) implements
+/// [`NetworkBehaviour`]. Multiple implementations of [`NetworkBehaviour`] can be composed into a
+/// hierarchy of [`NetworkBehaviour`]s where parent implementations delegate to child
+/// implementations. Finally the root of the [`NetworkBehaviour`] hierarchy is passed to
+/// [`Swarm`](crate::Swarm) where it can then control the behaviour of the local node on a libp2p
+/// network.
 ///
-/// Crate users can implement this trait by using the the `#[derive(NetworkBehaviour)]`
+/// # Hierarchy of [`NetworkBehaviour`]
+///
+/// To compose multiple [`NetworkBehaviour`] implementations into a single [`NetworkBehaviour`]
+/// implementation, potentially building a multi-level hierarchy of [`NetworkBehaviour`]s, one can
```

Comment:
Is there a real-world example of a multi-level hierarchy? Not just the one level, but something that has 2 or more levels.

Reply:
Yes. You could imagine using something like `libp2p-request-response` to make a custom request-response protocol but at the same time compose that with `libp2p-ping` into your root behaviour.

Reply:
To add an example to @thomaseizinger comment, say you would use both `libp2p-ping` and [`libp2p-autonat`](https://github.com/libp2p/rust-libp2p/pull/2262/), your `NetworkBehaviour` hierarchy would be:

- `MyRootBehaviour`
  - `AutoNat`
    - `RequestResponse`
  - `Ping`